tools/StatelessExecution/SetupCli.cs

```diff
@@ -0,0 +1,177 @@
+            (ulong)suggestedBlockForRpc.Timestamp,
+            suggestedBlockForRpc.ExtraData,
+            suggestedBlockForRpc.BlobGasUsed,
+            suggestedBlockForRpc.ExcessBlobGas,
+            suggestedBlockForRpc.ParentBeaconBlockRoot,
+            suggestedBlockForRpc.RequestsHash)
+        {
+            StateRoot = suggestedBlockForRpc.StateRoot,
+            TxRoot = suggestedBlockForRpc.TransactionsRoot,
+            ReceiptsRoot = suggestedBlockForRpc.ReceiptsRoot,
+            Bloom = suggestedBlockForRpc.LogsBloom,
+            GasUsed = suggestedBlockForRpc.GasUsed,
+            MixHash = suggestedBlockForRpc.MixHash,
+            BaseFeePerGas = suggestedBlockForRpc.BaseFeePerGas!.Value,
+            WithdrawalsRoot = suggestedBlockForRpc.WithdrawalsRoot,
+            ParentBeaconBlockRoot = suggestedBlockForRpc.ParentBeaconBlockRoot,
+            RequestsHash = suggestedBlockForRpc.RequestsHash,
+            BlobGasUsed = suggestedBlockForRpc.BlobGasUsed,
+            ExcessBlobGas = suggestedBlockForRpc.ExcessBlobGas,
+            Hash = suggestedBlockForRpc.Hash,
+        };
+
+        BlockHeader? baseBlock = witness.DecodedHeaders.FirstOrDefault(h => h.Hash == suggestedBlockHeader.ParentHash);
+        if (baseBlock is null)
+        {
+            throw new ArgumentException($"No base block header in witness. Expected header with hash {suggestedBlockHeader.ParentHash}");
+        }
+
+        var transactions = new Transaction[suggestedBlockForRpc.Transactions.Length];
+        for (int j = 0; j < transactions.Length; j++)
+        {
+            var tx = (JsonElement)suggestedBlockForRpc.Transactions[j];
+            transactions[j] = serializer.Deserialize<TransactionForRpc>(tx.GetRawText()).ToTransaction();
+        }
+
+        Block suggestedBlock = new Block(suggestedBlockHeader, transactions, [], suggestedBlockForRpc.Withdrawals);
+        return (witness, suggestedBlock, baseBlock);
+    }
+
+    private static bool ProcessBlock(BlockHeader baseBlock, Witness witness, Block suggestedBlock,
+        ISpecProvider specProvider, ILogManager logManager)
+    {
+        ILogger logger = logManager.GetClassLogger();
+        StatelessBlockProcessingEnv blockProcessingEnv =
+            new(witness, specProvider, Always.Valid, logManager);
+
+        IBlockProcessor blockProcessor = blockProcessingEnv.BlockProcessor;
+
+        using var scope = blockProcessingEnv.WorldState.BeginScope(baseBlock);
+        try
+        {
+            (Block processed, TxReceipt[] _) = blockProcessor.ProcessOne(suggestedBlock,
+                ProcessingOptions.ReadOnlyChain, NullBlockTracer.Instance, specProvider.GetSpec(suggestedBlock.Header));
+
+            if (processed.Hash != suggestedBlock.Hash)
+            {
+                logger.Info("❌  Block processing failed - Hash mismatch");
+                logger.Info($"    Expected: {suggestedBlock.Hash}");
+                logger.Info($"    Actual:   {processed.Hash}");
```

Comment:
Are those logger calls being executed under zk-evm?
Maybe worth having them removable by JIT?

Reply:
Probably not. I don't know if zkvms support logging, maybe some do. We can adjust it later, maybe we'll need some type of special logger for that.